samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala

```diff
@@ -483,24 +490,14 @@ object SamzaContainer extends Logging {
 
       val taskName = taskModel.getTaskName
 
-      val taskObj = if (taskFactory != null) {
-        debug("Using task factory to create task instance")
-        taskFactory match {
+      debug("Using task factory to create task instance")
+      val task =
+        finalTaskFactory match {
```

Comment:
TaskFactories only returns an Object, we will still need to do if {} else {} based on whether it is an instance of AsyncStreamTaskFactory or StreamTaskFactory. The code would end up the same.